[feature/loginService > develop][#3] Login-Service 구현

shin-mallang opened the pull request:
구현한 기능

- Spring Security 추가
- 회원가입 기능 구현
- Json으로 로그인하는 필터 구현 
- JWT를 사용하여 인증하는 필터 구현


TODO 

- JWT에 member의 pk(Id)만 가지도 있도록 하여, 이대로 배포할 경우 보안상 문제가 생길 수 있음 
    - 관련 method : JwtService의 createAccessToken
  
- SecurituConfig 의 .antMatchers("/login", "/signUp").permitAll()
    - url을 이렇게 관리할지, 필드로 따로 빼서 관리할지 고민

- JwtAuthenticationFilter가 Filter를 구현하는 경우 두번 작동하는 문제
    - OncePerRequestFilter를 상속받도록 하여 고쳤으나, Filter를 구현할 경우 왜 두번 작동하는지 모르겠음
    - 위의 원인은 해당 블로그를 참고 [https://pgnt.tistory.com/102, https://justforchangesake.wordpress.com/2014/05/07/spring-mvc-request-life-cycle/]

close #3

shin-mallang committed:
[chore] spring-security 의존성 추가

- build.gradle에 spring-security 의존성 추가
- build.gradle에 java-jwt 의존성 추가
- application-jwt.yml 파일 사용을 위한 application.yml 파일 수정
- .gitignore 파일에 application-jwt.yml 추가

shin-mallang committed:
[feat] SecurityConfig 파일 기본 설정

- formLogin 비활성화
- httpBasic 비활성화
- csrf 비활성화
- session stateless 하게 설정
- login, signup 요청에 대하여 권한 인증 체크 X
- 나머지 요청들에 대해 인증 체크

shin-mallang committed:
[fix] BaseException 이 RuntimeException 을 상속받도록 수정

- 기존 BaseException 은 아무 예외도 상속받지 않은 상태였음
이를 RunTimeException을 상속받도록 처리

shin-mallang committed:
[fix] build.gradle 오타 수정

- 기존 :  testImplementation 'org.springframework.security:spring-security-test'c

- 변경사항 : 마지막 c 제거

shin-mallang committed:
[feat] MemberRepository 추가

- JpaRepository 상속
- Username을 통해 Member를 조회하는 메서드 추가

shin-mallang committed:
[feat] MemberService 인터페이스와 관련 DTO 추가

- UserDetailsService 상속받아 해당 인터페이스 구현체가 loadUserByUsername 을 통해 인증을 진행하도록 설정.

- create 메서드 추가
  회원가입 기능 수행

shin-mallang committed:
[feat] MemberServiceImpl, MemberException 추가

- MemberService를 구현한 MemberServiceImpl 추가
- Member 관련 예외를 처리할 MemberException 추가

shin-mallang committed:
[feat] SecurityConfig 에 PasswordEncoder 추가

- 회원가입 시 비밀번호를 암호화 해줄 Spring Serurity의 PasswordEncoder를 SecurityConfig 에 @Bean 으로 등록

shin-mallang committed:
[test] MemberService 에 대한 테스트코드 추가

- 회원가입
    - 회원가입 성공
    - 회원가입 실패 (원인 : 아이디 중복)

shin-mallang committed:
[feat] Jwt 관련 기능 추가

- JwtService 추가
    - AccessToken 발급하는 기능
    - JWT 로부터 MemberId 추출하는 기능
    - 발급된 AccessToken 을 Header 에 추가하여 반환하는 기능
    - Http 요청으로부터 AccessToken 추출하는 기능
    - 토큰이 유효한지 검사하는 기능

- JwtToken 클래스 추가
    - JWT를 일반 String 이 아닌 객체로 관리하기 위해 생성

shin-mallang committed:
[fix] MemberServiceImpl 코드 수정

- loadUserByUsername 수정
    - User 객체에 username 으로 member의 id 를 전달

shin-mallang committed:
[feat] RestfulLoginFilter 추가

- RestfulLoginFilter 구현
    - 로그인 시 username과 password를 json으로 받아서 로그인을 처리하고 인증 정보를 저장하도록 구현

shin-mallang committed:
[feat] RestfulAuthenticationSuccessHandler 추가

- RestfulAuthenticationSuccessHandler 구현
    - 로그인 성공 시 후처리 역할을 구현하도록 AuthenticationSuccessHandler 를 구현
    - 로그인에 성공하면 JWT 발급

shin-mallang committed:
[feat] JwtAuthenticationFilter 추가

- JwtAuthenticationFilter 구현
    - AccessToken 을 통해 인증을 진행하는 filter

shin-mallang committed:
[feat] SecurityConfig 에 인증 관련 filter 추가

- JwtAuthenticationFilter 등록
- UsernamePasswordAuthenticationFilter 등록
- RestfulAuthenticationSuccessHandler 등록

shin-mallang committed:
[feat] MemberController 추가

- 회원가입 요청 처리
    - 회원가입 성공 시 201 상태코드 반환

shin-mallang committed:
[fix] JwtService 의 isValid 수정

- JwtService의 isValid 수정
    - JwtService의 isValid가 JwtToken 을 매개변수로 받아오도록 수정

- JwtToken 에 isValid 메서드 추가
    - 유효성 여부를 판단하는 것은 JwtToken의 역할이라 판단하여 이렇게 구현함

shin-mallang committed:
[fix] JwtService 의 extractToken 수정

- JwtService의 extractToken 수정
    - JwtService의 extractToken이 JwtToken을 반환하도록 수정

shin-mallang committed:
[test] JwtService 테스트코드 작성

- 토큰 유효성 검사
- 토큰 응답의 Header로 전송
- AccessToken 발급
- 토큰으로부터 회원정보 추출
- 요청으로부터 AccessToken 추출
- 요청 -> 토큰 추출 -> 회원정보 추출

shin-mallang committed:
[style] import문 정리

- 필요없는 import문 제거

shin-mallang committed:
[fix] JwtAuthenticationFilter가 OncePerRequestFilter를 상속받도록 수정

- 기존 Filter를 구현하였을 때 필터가 두번 작동되는 오류가 있었음
    - 원인은 아직 모르겠음
    - 한번만 작동하도록 OncePerRequestFilter를 상속받아 구현함

shin-mallang committed:
[test] 인증 관련 테스트코드 작성

- 로그인시 AccessToken 발급 테스트
- AccessToken을 가지고 접근 테스트

shin-mallang committed:
[style] JwtAuthenticationFilter안쓰는 필드 제거, TODO 추가

- NO_CHECK_URL 제거

ProtoSeo commented on src/main/java/com/footprint/member/controller/SignUpRequest.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Created by ShinD on 2022/05/18.
+ */
+public record SignUpRequest (String username, String password, String nickName) {
```
저희가 `Member` 도메인에서 별명을 `nickname`으로 사용하고 있기도 하고, nickname 자체가 하나의 명사여서 nickName보다는 nickname으로 사용하는 것이 어떨까요?

shin-mallang replied:
Camel case가 습관이 되어서 nickName으로 적은 것 같습니다. 수정하겠습니다.

ProtoSeo commented on src/test/java/com/footprint/auth/AuthTest.java:
```diff
@@ -0,0 +1,169 @@
+	public void failLoginByWrongPassword() throws Exception {
+		//given
+		Map<String, String> authMap = signUp();
+		authMap.put("password", "1");
```
잘못된 비밀번호로 1을 주는 것보다 명확하게 `"wrong-password"` 이렇게 주는 것은 어떨까요?

shin-mallang replied:
`static final String WRONG_PASSWORD = "wrong-password"` 은 어떨까요?

ProtoSeo replied:
좋습니당!

ProtoSeo reviewed (changes requested):
확인해주시면 감사하겠습니다!

ProtoSeo commented on src/main/java/com/footprint/config/SecurityConfig.java:
```diff
@@ -0,0 +1,79 @@
+
+			.and()
+			.authorizeRequests()
+			.antMatchers("/login", "/signUp").permitAll()//TODO 따로 빼서 관리하는 것이 나을지, 아니면 이대로 사용할지
```
따로 빼서 관리한다는 것이 어떤 방식으로 하는 건지 궁금해요!

shin-mallang replied:
변수로 추출해서 `List<String> permitAllUrls =  Arrays.asList("/login", "/signUp')` 식으로 관리하는것을 생각해보았습니다.

ProtoSeo replied:
아항! 이건 뭐가 더 좋은 방법일지 좀 더 생각해보면 좋을 것 같아요! 지금은 이렇게 해도 괜찮을 것 같습니다~

ProtoSeo commented on src/main/java/com/footprint/member/exception/MemberExceptionType.java:
```diff
@@ -0,0 +1,41 @@
+
+
+
+	private final int errorCode;
```
저 errorCode에 대해서 생각해봤는데 `String` 으로 에러코드를 주면 어떨까요? Member 관련 에러면 `M101`, `M102`.. 댓글 관련이면 `C101`, `C102`.. 이런 식으로,, 이건 나중에 더 이야기 해보면 좋을 것 같아요! 저는 일단은 HttpStatus의 value값으로 구현했습니다.

shin-mallang replied:
좋은 생각인 것 같습니다.
수정은 지금 바로 할까요?

ProtoSeo replied:
아뇨! 나중에 프론트랑 errorCode 관련된 이야기 할 때? 그 때 같이 수정하면 될것 같아요!

ProtoSeo commented on src/main/java/com/footprint/member/exception/MemberExceptionType.java:
```diff
@@ -0,0 +1,41 @@
+
+
+
+	private final int errorCode;
```
errorCode관련해서 생각해봤는데, String으로 관리하는 것은 어떨까요? Member 관련 예외면, `M101`, `M102` ... 그리고 댓글 관련이면 `C101`, `C102`... 이런식으로 말이죠! 이건 나중에 더 얘기 나눠보면 좋을 것 같습니다! 저는 지금은 HttpStatus value값 가져와서 예외처리 하도록 구현했습니당

shin-mallang committed:
[refactor] 회원 DTO에서 nickName -> nickname으로 변경

- CreateMemberDto의 nickName을 nickname으로 변경
- SignUpRequest nickName을 nickname으로 변경

shin-mallang committed:
[refactor] AuthTest 변수명 리팩토링

- 잘못된 Username과 password, 유효하지 않은 AccessToken 을 변수로 빼서 관리

shin-mallang commented:
수정 완료했습니다! :D
